File: codejam/2008/amer_semifinal/code_sequence/solution.hs
import Data.List (intersperse, nub, partition, sort)

infixr 0 <|
(<|) f = f
(<||) f g x = f <| g x

modp :: Int -> Int
modp = flip mod 10007

bifurcate :: [Int] -> ([Int], [Int])
bifurcate = (fmap (,) (map snd <|| fst) <*> (map snd <|| snd))
    <|| partition (even <|| fst)
    <|| zip [0..]

previousDiff :: [Int] -> Maybe Int
previousDiff diffs
    | length evenDiffs <= 1 = Nothing
    | varyingEvenDiffs = Just <| head oddDiffs
    | otherwise = previousDiff oddDiffs
    where
        (evenDiffs, oddDiffs) = bifurcate diffs
        varyingEvenDiffs = (>1) <| length <| nub evenDiffs

nextTerm :: [Int] -> Maybe Int
nextTerm vs = (fmap <| modp <|| (+ last vs))
    <| previousDiff
    <| reverse
    <| map (modp <|| uncurry (-))
    <| (tail >>= zip) vs

format :: Maybe Int -> String
format Nothing  = "UNKNOWN"
format (Just v) = show v

solve :: IO String
solve = getLine >> fmap (map (read :: String -> Int) <|| words) getLine
    >>= return
        <|| format
        <|| nextTerm

main :: IO ()
main = (readLn :: IO Int)
    >>= sequence <|| flip replicate solve
    >>= putStrLn
        <|| concat
        <|| intersperse "\n"
        <|| map (uncurry (++))
        <|| zip (map (("Case #" ++) <|| (++ ": ") <|| show) [1..])

File: codejam/2008/amer_semifinal/code_sequence/solution.rs
lib::run!();

const M: usize = 10007usize;

fn read() -> Vec<usize> {
    lib::input!(usize as _l);
    lib::input!(Vec<usize> as sequence);
    sequence
}

fn solve(mut seq: Vec<usize>) -> String {
    seq.reverse();
    let mut diffs = Vec::new();
    for i in 0..seq.len() - 1 {
        diffs.push((seq[i] + (M - seq[i + 1])) % M);
    }
    if let Some(d) = previous_diff(diffs) {
        format!("{}", (seq[0] + d) % M)
    } else {
        "UNKNOWN".to_string()
    }
}

fn previous_diff(diffs: Vec<usize>) -> Option<usize> {
    if diffs.len() < 3 {
        return None;
    }
    let mut odd_diffs: Vec<usize> = Vec::new();
    for i in (1..diffs.len()).step_by(2) {
        odd_diffs.push(diffs[i]);
    }
    let mut varying_even_diffs = false;
    for i in (0..diffs.len() - 2).step_by(2) {
        varying_even_diffs |= diffs[i] != diffs[i + 2];
    }
    if varying_even_diffs {
        Some(odd_diffs[0])
    } else {
        previous_diff(odd_diffs)
    }
}
